Expose Ocean Variables

kdorheim opened the pull request:
**Describe development**

This PR is related to #436,  we want to expose the R bindings for the ocean variables so that we can fetch the variables from the Hector core. I did run into some issues where a unitval was not properly defined and it turns out some of the variables are not returned as time series, which variables do we want to return as a complete-time series. 


**Type of development** 

* MINOR: adding a capability for the R wrapper

**Checklist**
- [x] [Roxygen](https://roxygen2.r-lib.org/) where applicable but also thorough and coherent inline documentation 
- [x] [Unit tests](https://r-pkgs.org/tests.html) when applicable
- [x] Complies with the [Hector code style guide](https://jgcri.github.io/hector/articles/manual/StyleGuide.html)
- [x] The PR passes automated testing suite

kdorheim commented on src/ocean_component.cpp:
```diff
@@ -611,6 +630,7 @@ void OceanComponent::record_state(double time)
     annualflux_sumLL_ts.set(time, annualflux_sumLL);
     lastflux_annualized_ts.set(time, lastflux_annualized);
+    C_IO_ts.set(time, inter.get_carbon());
 
```
@bpbond 

Here I turned one of the tvectors into a series so that I could return them as ouptut using fectchvars, is this how we want to do it? Or is there some what to return the series of values from the vector?

bpbond replied:
This is unfortunate but seems like the right way to go.

kdorheim replied:
yea okay messy but good to know

kdorheim commented on src/ocean_component.cpp:
```diff
@@ -81,6 +81,18 @@ void OceanComponent::init( Core* coreptr ) {
+    core->registerCapability( D_TEMP_HL, getComponentName() );
+    core->registerCapability( D_TEMP_LL, getComponentName() );
+    core->registerCapability( D_CO3_HL, getComponentName() );
+    core->registerCapability( D_CO3_LL, getComponentName() );
```
@bpbond  I assume that we want to return ph, the ocean fluxes, pco2, dic, temp,  and co3 correct? For the high and low latitude, we only want to return the surface level values, we don't want to sum over the intermediate and deep occeans correct?

bpbond replied:
That seems like a good place to start. Yes, just surface values; there's no chemistry in the intermediate and deep boxes.

bpbond commented:
Will this close #436 ? Also, are your changes done—should I review full changes?

kdorheim commented:
> Will this close #436 ? Also, are your changes done—should I review full changes?

Yes it will close 436, I have a few more things to do, I have to add the other time series as outputs. And then I figured I would probably add these variables to unit test suite, I figured old new tests for ocean variables would be pretty helpful as part of the doeclim update.

bpbond commented:
@kdorheim Sounds good. I'd say proceed, looks good. Thanks!

kdorheim commented:
Alright, so I have exposed R bindings for the output variables as well as some parameters, updated documentation, and added some unit tests. 

Here are some plots of the ocean output variables for the RCP 4.5 scenario. Note that evaluating how these outputs compare to observations/other models is a separate question that should/will be addressed elsewhere. 

# Ocean output variables 
![image](https://user-images.githubusercontent.com/27299759/116733417-1407d180-a9ba-11eb-99da-c93a4357de1f.png)


# The different ocean carbon pools 
![image](https://user-images.githubusercontent.com/27299759/116733430-1702c200-a9ba-11eb-9d8f-613e6a06959c.png)

bpbond commented on data-raw/generate-comp-data.R:
```diff
@@ -19,8 +19,8 @@ hector_version <- packageVersion("hector")
-vars_to_save <- c(ATMOSPHERIC_CO2(), GLOBAL_TEMP(), RF_TOTAL(), RF_CO2(), HEAT_FLUX())
-dates_to_save <- 1750:2300
+vars_to_save <- c(ATMOSPHERIC_CO2(), GLOBAL_TEMP(), RF_TOTAL(), RF_CO2(), HEAT_FLUX(), OCEAN_C(), PH_HL())
+dates_to_save <- 1850:1950
```
Why the date change?

kdorheim replied:
It is an arbitrary change but as we added more variables to the comparison data file I wanted to keep it smaller 🤷

bpbond commented on tests/testthat/test_ocean.R:
```diff
@@ -0,0 +1,118 @@
+
+
+test_that("Checking carbon pools", {
+
```
How much does it differ by? That is, is this a fundamental structural problem, or just a numerics issue? We wouldn't expect them to be _exactly_ equal.

kdorheim replied:
So it is different by 34743, which is larger than I would expect for a numerical issue 

```
Error: `total_ocean_pool_value` not equal to sum_carbon_pools$value.
51/51 mismatches (average diff: 34743)
[1] 594 - 35339 == -34745
[2] 594 - 35339 == -34744
[3] 595 - 35339 == -34744
[4] 595 - 35339 == -34744
[5] 595 - 35339 == -34744
```

kdorheim replied:
Okay looking into it more according to the Wiki output files page the output being returned with the 
`fetchvars(core = hc, dates = t_dates, vars = OCEAN_C())` reports the incorrect units. 
```
ocean | ocean_c | Pg C | total ocean carbon pool
```

kdorheim replied:
@bpbond I think I found the problem there is still a difference between the total ocean pool and the sum of the individual ocean pools 

```
Error: total_ocean_pool$value not equal to sum_carbon_pools$value.
50/51 mismatches (average diff: 4.03)
[1] 35345 - 35339 == 6.75
[2] 35345 - 35339 == 6.70
[3] 35345 - 35339 == 6.64
[4] 35345 - 35339 == 6.58
```
Considering that this is a relatively small difference ` 0.0114019`% @bpbond do you consider this os be a trivial numerical difference we can tolerate in our tests?

kdorheim replied:
@bpbond and I discussed that this is fine but want to stage this as an issue to look into in the future.

bpbond replied:
--> #440

bpbond commented on tests/testthat/test_ocean.R:
```diff
@@ -0,0 +1,118 @@
+        # What is the absolute difference between the mean ocean variables?
+        error_threshold <- 1e-10
+        diff <- abs(new_mean$value - default_mean$value)
+        expect_true(all(diff > error_threshold), info = cat('problem with resetting ', p, '\n'))
```
This makes sense to me.

bpbond reviewed (approved):
Looks good overall. See my question below about the failing test.

bpbond commented:
Look at those lovely green checks starting to pop up @kdorheim

kdorheim commented:
Sigh I forgot to update the comp data for the old-new test after I fixed the total pool that was being retired 🤦‍♀️

bpbond commented:
@kdorheim OK to merge?

kdorheim commented:
Okay it somewhat annoys me that the travis CI is still hanging around even though it isn't running anymore but @bpbond I think it is ready to merge!

bpbond commented:
Me too re annoyance. I couldn't figure out how to kill that permanently but will try again at some point. OK, thanks.
